src/app-layout/visual-refresh/hooks/use-dynamic-overlap.ts

```diff
@@ -0,0 +1,40 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { useContext, useLayoutEffect } from 'react';
+
+import { AppLayoutContext } from '../context';
+import { useContainerQuery } from '../../../internal/hooks/container-queries';
+
+export interface UseDynamicOverlapProps {
+  /**
+   * Disables hook if not needed. By default, `false`.
+   */
+  disabled?: boolean;
+}
+
+/**
+ * Observes the height of an element referenced by the returning ref and sets the value as overlapping
+ * height for the surrounding AppLayout component.
+ * @param props.disabled disables hook if not applicable
+ * @returns ref to be measured as overlapping height
+ */
+export function useDynamicOverlap(props?: UseDynamicOverlapProps) {
+  const disabled = props?.disabled ?? false;
+  const { setDynamicOverlapHeight } = useContext(AppLayoutContext);
+  const [overlapContainerQuery, overlapElementRef] = useContainerQuery(rect => rect.height);
+
+  const hasOverlap = !disabled && overlapContainerQuery !== null;
+  useLayoutEffect(
+    function handleDynamicOverlapHeight() {
+      setDynamicOverlapHeight(hasOverlap ? overlapContainerQuery : 0);
```

Comment:
Technically this will work because the `dynamicOverlapHeight` is instantiated as a zero value in the `AppLayout` context and the logic for overriding the custom property is:

`dynamicOverlapHeight > 0 && { [customCssProps.overlapHeight]: '${dynamicOverlapHeight}px' })`

So, if the `disabled` property in this hook is `true` then we will be calling the setter in the context to update the value to zero, which it likely already is. I'm wondering if we even need this `disabled` property. Ideally the component that implements the hook will only use it under circumstances where it _isn't_ disabled which would make this property irrelevant.

Is there a use case for the `disabled` property that I'm missing? Similarly, if internally the `overlapContainerQuery` is `null` we should not even call the setter, I think.

Reply:
If we don't call the setter when `disabled` or `overlapContainerQuery` flips, aren't we missing update to the "reactive" properties like we do right now with the lack of a cleanup function? Keep in mind that this function is not just called on first render of the overlap component but whenever state or properties change.

Reply:
Continuing this conversation in the other thread...